segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,378 @@
+            timeout) {
+            return CompletableFuture.supplyAsync(() -> syncConcat(targetHandle, offset, sourceSegment, timeout),
+                    executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+            return CompletableFuture.supplyAsync(() -> syncDelete(handle, timeout), executor);
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+
+    }
+
+    //endregion
+
+    //region private sync implementation
+
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+
+        if (!Files.exists(path)) {
+            throw new CompletionException(new StreamSegmentNotExistsException(streamSegmentName));
+        }
+
+        FSSegmentHandle retHandle = FSSegmentHandle.getReadHandle(streamSegmentName);
+        return retHandle;
+    }
+
+
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length, Duration
+            timeout) {
+
+        Path path = Paths.get(config.getNfsRoot(), handle.getSegmentName());
+
+        if (!Files.exists(path)) {
+            throw new CompletionException(new StreamSegmentNotExistsException(handle.getSegmentName(), null));
+        }
+
+        try {
+            if (Files.size(path) < offset) {
+                log.warn("Read called on segment {} at offset {}. The offset is beyond the current size of the file.",
+                        handle.getSegmentName(), offset);
+                throw new CompletionException(new ArrayIndexOutOfBoundsException());
+            }
+        } catch (IOException ioe) {
+            throw new CompletionException(ioe);
+        }
+
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) {
+            int bytesRead = channel.read(ByteBuffer.wrap(buffer, bufferOffset, length), offset);
+            return bytesRead;
+        } catch (IOException | IndexOutOfBoundsException e) {
+            if (e instanceof IndexOutOfBoundsException) {
+                throw new CompletionException(new ArrayIndexOutOfBoundsException(e.getMessage()));
+            } else {
+                throw new CompletionException(e);
+            }
+        }
+    }
+
+    private SegmentProperties syncGetStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        try {
+            PosixFileAttributes attrs = Files.readAttributes(Paths.get(config.getNfsRoot(), streamSegmentName),
+                    PosixFileAttributes.class);
+            StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName, attrs.size(),
+                    !(attrs.permissions().contains(OWNER_WRITE)), false,
+                    new ImmutableDate(attrs.creationTime().toMillis()));
+
+            return information;
+        } catch (IOException ioe) {
```

Comment:
Please translate ALL exceptions into Storage Exceptions. Here and elsewhere in the file.